core/src/main/java/org/elasticsearch/index/percolator/QueryMetadataService.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.index.percolator;
+
+import org.apache.lucene.document.Field;
+import org.apache.lucene.document.FieldType;
+import org.apache.lucene.index.*;
+import org.apache.lucene.queries.TermsQuery;
+import org.apache.lucene.search.*;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.index.mapper.ParseContext;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class QueryMetadataService {
+
+    public static String QUERY_METADATA = "_query_metadata_";
+    public static String QUERY_METADATA_FIELD_PREFIX = QUERY_METADATA + "field_";
+    public static String QUERY_METADATA_FIELD_UNKNOWN = QUERY_METADATA + "unknown";
+    public static FieldType QUERY_METADATA_FIELD_TYPE = new FieldType();
+
+    static {
+        QUERY_METADATA_FIELD_TYPE.setTokenized(false);
+        QUERY_METADATA_FIELD_TYPE.setIndexOptions(IndexOptions.DOCS);
+        QUERY_METADATA_FIELD_TYPE.freeze();
+    }
+
+    public void extractQueryMetadata(Query query, ParseContext.Document document) {
+        List<Term> queryTerms = new ArrayList<>();
+        extractQueryMetadata(query, queryTerms);
+        for (Term term : queryTerms) {
+            document.add(new Field(term.field(), term.bytes(), QUERY_METADATA_FIELD_TYPE));
+        }
+    }
+
+    public void extractQueryMetadata(Query query, List<Term> queryTerms) {
```

Comment:
I think we could make it a better API by changing the signature to `List<Term> extractQueryTerms(Query query)` and throwing an exception if they can't be extracted. Then it would be the responsibility of the caller to add the `QUERY_METADATA_FIELD_PREFIX` prefix to field names?